### ISSUE TITLE
Adding UTRs to transcripts in the reference update mode

### ISSUE THREAD

**Issue by lucventurini** [EI-CoreBioinformatics/mikado#208]
In #207, @fmarletaz asked me whether we have a way to add UTRs to transcripts derived from eg. Augustus. This made me realise that currently this is not the case, **not even in reference update mode**.

The problem here is that Mikado explicitly disallows having transcripts that are identical inside a locus (ie.: the `=` and `n` [class codes](https://mikado.readthedocs.io/en/latest/Usage/Compare.html?highlight=class%20code#class-codes) are explicitly **forbidden** as valid ASE class codes in `pick`). This means that, in the not-so-hypothetical scenario in which we have a perfectly supported *ab initio* transcript lacking UTR, and its companion cDNA/assembly from RNASeq which would add the UTR itself, **we will not have a method to add it**.

Moreover, at the moment when doing the reference update, the only thing that happens is that Mikado will discard any locus that has no reference transcript. In subsequent stages, however, there is no guarantee that those transcripts will pass through at all.

There are ways to work around this (e.g.: avoid doing a proper mikado run and use instead `mikado compare` to drive an ancillary script to expand transcripts) but I would like to hear your opinions on the matter.

**Comment by lucventurini** [comment]
This issue will be worked on in [the branch issue-208](https://github.com/lucventurini/mikado/tree/issue-208).

**Comment by lucventurini** [comment]
In the latest commit for the branch, I started working on this. Basically, in this version, whenever padding is activated normally "redundant" transcripts (`=`, `_`, `n`) will be considered valid ASEs. 

Furthermore, after padding, for any couple of identical transcripts (to the base, ie nF1 == 100, given that we are talking about padded transcripts), only one will be kept (reference transcripts and higher score transcripts first). 

The problem with the approach is that implemented like this, we risk losing bunches of ASEs as Mikado will integrate all minor variants of transcripts into the locus ... and there is a hard limit on the number of ASEs. For the moment, the changes will be quarantined on this branch.

I think, in general, that they would be useful as they allow to properly integrate RNASeq and *ab initio* prediction, with UTR expansion and all--even though this is not ready for prime time.

**Comment by Tierhon** [comment]
This sounds great. I was just looking for something like this. Presently, I have been using Mikado to merge different transcript evidence, and then used that further in Funannotate together with mapped proteins (with GenomeThreader) and let Funannotate control training of Augustus (funannotate predict). However, I do believe that I have lost UTRs after this. Funannotate has its own method of updating the annotation using Kallisto to find the best transcripts and PASA and EVM to add the UTRs. This is quite cumbersome, and I have not gotten it to work properly. 

If I can use Mikado to add UTRs to the annotation from Funannotate, that would be perfect. Is it usable yet? 

Thank you. 

Ole

**Comment by lucventurini** [comment]
Hi Ole, it should be. I am currently on holiday, but next week I can write out the detailed instructions (I need to do it anyway for the documentation). 

It's not perfect or well tested yet so I would really appreciate the third party testing!

**Comment by swarbred** [comment]
Hi @lucventurini 

We definitely have a use for updating a reference annotation adding additional splice variants and extending UTRs. Marking a set of transcripts as "reference" allowed us to incorporate additional splice variants (SV), but we did this in a way that didn't require many changes to mikado (to save you work). So when we want to add SV to an exiting annotation we do the following 

1) split the reference annotation into two files, file (a) with just the primary models and file (b) with the non primary models.
2) in the list.txt file both file a and b are marked as reference (note I see in mikado configure --help we call this field always_keep rather than reference, which might confuse as it only really means alwasy keep for the prepare step). Any other models (which you want to add SV from) are not reference.
3) Also in the list.txt file you need to add a score against file a and file b, file a needs to be higher than b (e.g. 1000 and 500), this way you ensure that the same model is selected by pick as the primary, and that one of your reference models wins out at the pick stage
4) set min_score_perc in the config to 0.0 otherwise you will exclude all other SV (given the high base scores we gave each of our two reference transcript files)
5) run pick with the --only-reference-update enabled

This doesn't guarantee all reference transcripts are carried through but all the primary models (file a) should be present.

So this kind of works but it is convoluted.  The "reference" assignment was there to ensure models dont get filtered prior to the pick and there are times when you want to assign a set of transcripts as reference but not use the --only-reference-update function i.e. if you want to update and add to a set of models. 

On the UTR update 

"I amended this so that this automatic addition would happen only when executing a "reference-update"

That might be ok but there might be situations where you either want to do this without --only-reference-update being enabled or not do when that is enabled. I would consider a separate flag for the UTR behaviour  (and good documentation to explain all this).

"The problem with the approach is that implemented like this, we risk losing bunches of ASEs as Mikado will integrate all minor variants of transcripts into the locus ... and there is a hard limit on the number of ASEs."

I don't really follow this, shouldn't the limit on ASE be applied after your nF1 == 100 filter.

Making some minor tweeks to allow some form of UTR extension is useful, however to be honest there would probably be a better way to do the SV addition and UTR extension of an existing annotation i.e. a separate mikado tool but obviously thats a lot more work.

**Comment by lucventurini** [comment]
> I don't really follow this, shouldn't the limit on ASE be applied after your nF1 == 100 filter.

No, as the logic is written, the maximum number of isoforms is used as a cutoff **before** we even do the padding. It is basically a switch that says to Mikado "just stop adding" once a critical number of isoforms has been reached. It is **not** the most elegant solution, hence why I suggested maybe removing it.

> Making some minor tweeks to allow some form of UTR extension is useful, however to be honest there would probably be a better way to do the SV addition and UTR extension of an existing annotation i.e. a separate mikado tool but obviously thats a lot more work.

One that could still be done with the current infrastructure of the program, but yes, it would probably take a couple of weeks of dedicated programming to get it properly right. 

What would your preferred course of action with the changes I put in this branch? There are some (regarding making the construction of Mikado transcript graphs faster) that I think should go into the main branch no matter what, but the rest is a bit more of a grey area.

**Comment by lucventurini** [comment]
Hi @swarbred, @gemygk:

- I have merged most of the changes back into master. I have tested extensively and they seem to function.
- I am now testing (see 6c350ec) a change in the definition of alternative splicing which will allow to pad, remove redundancies, and still keep a meaningful number of maximum isoforms. See the commit for the changes. The *caveat* is that the threshold for keeping or discarding a transcript from the final gene **is now calculated when all candidates are present, only once**. Before, the threshold was recalculated every time (see the commit for details).

The *caveat* means that there might be instances where the percentage threshold is not respected, in the final version of the locus.

**Comment by lucventurini** [comment]
To clarify, Mikado does *not* ensure that the primary transcript has a stable score during the AS selection stage.
To the contrary, both with the previous method and the current, there are instances where the primary transcript gets superseded by other transcripts in the locus for the final scoring. This is due to the fact that of course there are many metrics that hugely depend on the transcripts that are present in the locus at a given time (e.g. `proportion_verified_introns_in_locus`) and which therefore will vary as we add or remove transcripts from a locus.
Even for stable metrics, the associated score will change (e.g. for scoring on the transcript with the greatest number of transcripts).
Calculating all possible combinations would probably be prohibitive, therefore Mikado uses as criterion for the addition only the initial scores.

**Comment by swarbred** [comment]
@lucventurini just to comment while I have this in my head if we want to "fully" enable updates to reference annotation ie new splice variants and UTR extensions then we might need to look at allowing some non-valid models

This ties in with https://github.com/EI-CoreBioinformatics/mikado/issues/226

i.e. if you wanted to update say the downloaded human ensembl gene file you would find many models that will not meet mikados validation i.e. pseudogenes with annotated CDS containing stops , annotation based on haplotypes again containing stops or other features annotated in ways that  mikado would not except. 

Currently mikado prepare would clean these up (which is obviously desirable in a normal use case and I assume still true even if the models are marked as reference) but in the context of wanting to update an existing annotation probably not desirable.

So potentially you could have a lenient mode that for prepare, stats, pick etc that passes through these models unchanged. For the pick I guess the working solution would be to ignore any loci containing these transcripts i.e. no updates are made i.e. it's a set of loci that for pick purposes and the generation of metrics are simply ignored.

I'm not suggesting we do this as I realise it's likely not straight forward but just commenting as you will see more issues like issues/226 if people use mikado with existing annotation.

**Comment by lucventurini** [comment]
Hi @swarbred ,
                           yes, I get your point.
Regarding the issue in #226, actually, paradoxically `mikado prepare` would not have a problem with that ... at that stage the program only collects exon/CDS/UTR features. So a model like the one causing the issue there (ie having only exons and pseudogene, but no transcript, defined) would probably pass through fine.

**Comment by Tierhon** [comment]
Hi @lucventurini. 

I see that you've changed the milestone to 2.1 a month ago or so. I guess this means that there is no straight-forward way of just adding UTRs to existing gene models (and only to valid models for instance, ignoring all non-valid).

Thank you.

**Comment by lucventurini** [comment]
I think this is now completely integrated in Mikado 2, due to progress on other issues.
